Fix zk-luhmann-id sorting

jgru opened the pull request:
Dear Grant, 

first of all thanks for providing `zk`. While working with this package, I stumbled over an issue with the sorting of the "Luhmann IDs". Previously, the IDs needed to include a space before the closing brace, which is fixed with the modification of the regex to match the Luhmann ID. In order to do this and stay customizable, I introduced `zk-luhmann-id-format` via `defcustom`

Even though that this is not "the officially packaged" `zk`, I thought I "upstream" this modification to you. 
Thanks again for `zk`. 

Best regards,
jgru

localauthor commented:
Thanks for this! I'll give it a look. First, though, its a bit hard to see what the actual changes are, since the diff shows a lot of changes where there don't seem to be changes. Is it possible to push a commit that includes only the actual changes? Can y you see what I mean, for example, starting from line 52?

jgru committed:
Fix zk-luhmann-id sorting

Fix zk-luhmann-id sorting by modifying the regex and introducing a
variable for specifying the ID-format to use.

jgru commented:
> Thanks for this! I'll give it a look. First, though, its a bit hard to see what the actual changes are, since the diff shows a lot of changes where there don't seem to be changes. Is it possible to push a commit that includes only the actual changes? Can y you see what I mean, for example, starting from line 52?

Thanks for getting back to me. Unfortunately, a save-hook performed some auto-formatting leading to the changes. I corrected that. Now, it's obvious what was changed. By using the incorporated regex `zk-luhmann-index-sort` works just fine without any trailing space before the last brace.

localauthor committed:
Add zk-word-count

minor init changes

localauthor commented:
Sorting seems to work properly in the minibuffer and zk-index. But completion at point isn't right, for some reason.

The functions `zk-luhmann-index-forward` and `-back` also don't work, since they expect a space before the bracket. They would have to be refactored to account for the change, if you're up for it.

Also, I'm not sure about the name `zk-luhmann-id-format`, since it's not really a format variable. Maybe `zk-luhmann-id-regexp` instead? And it should probably be a `defvar` instead of a `defcustom`, since most of the functions here won't use the variable, and therefore won't work if it gets customized.

localauthor commented:
Now that I'm thinking this through, I recall that I originally arrived at the current syntax (with brackets, commas, and a space at the end) when I was sorting and displaying these notes using the command-line rather than Emacs.

The Unix function `sort` sorted the notes how I wanted only when there was a space before the closing bracket. I notice now that the sort-by-name function in `dired` does the same:

with space:
![Screen Shot 2022-02-09 at 22 52 26](https://user-images.githubusercontent.com/78040295/153288070-a22d1a26-860d-4114-a80b-4c7399be22c6.jpg)

no space:
![Screen Shot 2022-02-09 at 22 52 40](https://user-images.githubusercontent.com/78040295/153288205-a9a6db5b-3ad8-47ce-b60e-30a7dd3f79b7.jpg)

I'm actually pretty inclined to have a syntax that will be handled as expected as often as possible without much (or ideally any) intervention, for example via regexp or a custom sorting function.

The `zk-luhmann-completion-at-point` case is a good example of why I'm inclined this way: whatever the sorting logic is under the hood of Emacs's completion-at-point, it seems to be the same as in `dired` and `sort`, such that the present syntax results in the expected sort order automatically, regardless of the completion-at-point interface (default, company, or corfu, for example).

If that comes only at the price of a single space, it seems like a price worth paying. :)

Interested to hear what you think, in light of this.

jgru committed:
Modify ID handling

jgru commented:
Thanks for diving into this. The sorting behavior you describe is standard alphanumeric sorting, I assume. 
`SPACE` has the ASCII value 32 which is lower than any digit and any other alphabetic character. 
Opposed to that `}` has the value 125 which is higher than any digit and alphabetic character. 
Therefore, using `SPACE` makes the sorting works "naturally" (in all modes applying this sort order).
This can be illustrated like this:
 ```
echo -e "{1,a }\n{1}" | xxd                                                  

00000000: 7b31 2c61 207d 0a7b 317d 0a              {1,a }.{1}.
```
Since I think that having a space before the closing delimiter is not nice, I would like to propose to use parentheses. 
Those are assigned ASCII values smaller than digits and alphabetic characters so that sorting (and completion) works. 
  
Example: 
```
cat test.txt 

(1,a,1) ipsum lorem
(1) Lorem ipsum
(1,a) foo bar
(1,b) bar foo
```
Result by sorting with Unix' `sort`:
```
sort -r test.txt

(1) Lorem ipsum
(1,b) bar foo
(1,a) foo bar
(1,a,1) ipsum lorem
```
I started to implement support for this ID-scheme. Imho it is working well. However, until now I didn't get `zk-luhmann-index-level` to work. I will ping you, if I got it right. 

What do you think of this approach?

Best regards,
jgru

localauthor commented:
Thanks for clarifying for me the underlying sorting logic. It makes things much less opaque than they were.

As for figuring out a viable syntax here, i’m hesitant about opting for parentheses, since, at least in my case, I use parentheses already in file names for literature notes, to denote year of publication.

I’ll play around with it and see if that’s a problem or if maybe there’s another alternative.

Thanks again for working on this!
On Feb 10, 2022, 09:41 +0200, Jan ***@***.***>, wrote:
> Reopened #1.
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.
> You are receiving this because you commented.Message ID: ***@***.***>

localauthor commented:
> As for figuring out a viable syntax here, i’m hesitant about opting for parentheses, since, at least in my case, I use parentheses already in file names for literature notes, to denote year of publication. 

It seems like this becomes an issue in `zk-luhmann-files`, since it lists all files with anything in parens. So maybe that function could be made more specific:

```elisp
(defun zk-luhmann-files ()
  "List notes with Luhmann-IDs."
  (zk--directory-files t (concat zk-id-regexp " " zk-luhmann-id-start-char)))
```

localauthor commented:
Just a heads up, I pushed some cleanup and a fix for `zk-luhman-index-back`, so you'll want to rebase this branch onto main.

jgru commented:
> >

Thanks for your heads up on this. I think, we could match the ID only positioned after the TS. 
Alternatively, one could force the presence of at least one `,` in the ID. What do you think about this?
Personally, I encode the publication year in the form `firstauthorYear` in bibliographic notes.

jgru committed:
Modify filtering for files with Luhmann IDs

Modify filtering for files with Luhmann IDs by taking the position of
the Luhmann ID into account.

jgru commented:
Hi, a quick heads up I change the file filtering.
BTW, what do you think of putting `zk-luhmann.el` into a development branch of `zk` or creating a separate repo for it? 
Now and then, I intend to add some other functionality. 

Best regards, 
jgru

localauthor commented:
Sorry, could you make sure the auto-formatting is off when you push commits? It's hard to read again.

I think a separate repo is a good idea. I'll set that up. Then maybe you can open a PR there, and just include a reference to this one?

jgru commented:
> when you push commits? It's hard to read again.

Sure. Sorry, that the one commit slipped through. 

> I think a separate repo is a good idea. I'll set that up. Then maybe you can open a PR there, and just include a reference to this one?

Great, let's do it that way! Then I'll contribute a clean PR.

localauthor commented:
Ok, new repo here:  https://github.com/localauthor/zk-luhmann

jgru commented:
> Ok, new repo here: https://github.com/localauthor/zk-luhmann

Thanks for the creation. I'll open a cleaned-up PR tomorrow.

localauthor commented:
I've put most of this, with some changes and cleanup, in the dev branch of that repo.

As far as I can tell, the functions in that branch work the same as in main. See what you think, and open a PR there with any changes.
